design/wait-for-additional-items.md

```diff
@@ -0,0 +1,239 @@
+
+## Goals
+
+- Enable Velero to ensure that Additional items returned by a restore
+  plugin's `Execute` func are ready before restoring the current item
+- Augment the RestoreItemAction plugin interface to allow the plugins
+  to determine when an additional item is ready, since doing so
+  requires knowledge specific to the resource type.
+
+## Background
+
+Because Velero does not wait after restoring additional items to
+restore the current item, in some cases the current item restore will
+fail if the additional items are not yet ready. Velero (and the
+`RestoreItemAction` plugins) need to implement this "wait until ready"
+functionality.
+
+## High-Level Design
+
+After each RestoreItemAction execute call (and following restore of
+any returned addional items) , we need to wait for these returned
+Additional Items to be ready before restoring the current item. In
+order to do this, we also need to extend the `RestoreItemActionExecuteOutput`
+struct to allow the plugin which returned additional items to
+determine whether they are ready.
+
+## Detailed Design
+
+### `restoreItem` Changes
+
+When each `RestoreItemAction` `Execute()` call returns, the
+`RestoreItemActionExecuteOutput` struct contains a slice of
+`AdditionalItems` which must be restored before this item can be
+restored. After restoring these items, Velero needs to be able to wait
+for them to be ready before moving on to the next item. Right after
+looping over the additional items at
+https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L960-L991
+
+we still have a reference to the additional items (`GroupResource` and
+namespaced name), as well as a reference to the `RestoreItemAction`
+plugin which required it.
+
+At this point, if the `RestoreItemActionExecuteOutput` includes a
+non-nil `AdditionalItemsReadyFunc` we need to call a func similar to
+`crdAvailable` which we will call `itemsAvailable`
+https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L623
+This func should also be defined within restore.go
+
+Instead of the one minute CRD timeout, we'll use
+`resourceTerminatingTimeout`, and instead of the
+`IsUnstructuredCRDReady` call, we'll call the returned
+`AdditionalItemsReadyFunc` passing in the same `AdditionalItems` slice
+as an argument (with items which failed to restore filtered out). If
+this func returns an error, then `itemsAvailable` will
+propagate the error, and `restoreItem` will handle it the same way it
+handles an error return on restoring an additional item. If the
+timeout is reached without ready returning true, velero will continue
+on to attempt restore of the current item.
+
+### `RestoreItemActionExecuteOutput` changes
+
+A new func field will be added to `RestoreItemActionExecuteOutput`. If
+`AdditionalItemsReadyFunc` is non-nil, then `restoreItem` will call
+`itemsAvailable` which will invoke the plugin func
+`AdditionalItemsReadyFunc` and wait until the func returns true or the
```

Comment:
Would it be simpler to have a restore level option of `--sync-restore-additional-items`, a new field in the `RestoreSpec`, which if set, we would call a `waitForAdditionalItems` function after finishing the `for` loop at https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1026-L1056
The `waitForAdditionalItems` would use the `wait.PollImmediate` API to wait for the additional items to be ready. The time out for this can also be made configurable either using another restore level option, say, `--sync-restore-additional-items-timeout`.

The simplicity coming from:
1. With a new release of Velero, after this is implemented, all RestoreItemActions will be able to take advantage of this feature without having to make any changes to their plugins.
1. Current `RestoreItemAction` would not require to be updated. Similar to the limitation of an earlier iteration of the design.
1. The proposed design here, would require a release of Velero, with the updated `RestoreItemActionOutput` API type before plugins are able to release a version using the new fields in the `RestoreItemActionOutput`
1. Implementation of the `AdditionalItemsReadyFunc` would be very identical across all RestoreItemAction plugins. So, instead of having the same implementation across many plugins, this function may be implemented as the `waitForAdditionalItems` and unexported function in the restore package of core Velero.

Reply:
@ashish-amarnath interesting thought but I have a question:

In the solution you proposed, how would `waitForAdditionalItems` be implemented in a generic way for all resources?

The concept of "Readiness" may be different for different resources (for example for a pod it may be the container ready status, for a secret, it might be the meager existence and for a custom resource it might be a custom field in the status), and we need the plugin for a particular resource to defined what Readiness means for the resource it needs to wait.

Reply:
@alaypatel07 Yes, `waitForAdditionalItems` is generic across all resources. This function would wait for the restored resource to be present in the response to the `Get` from the API server.
I interpreted "ready" as "existence". It appeared that was the race that issue #1350 was referring to, that, restored resources were not immediately available.

Reply:
@ashish-amarnath So the basic reason for the plugin-specific function was that in discussing it with @nrb we weren't sure whether it was appropriate to attempt "is it ready?" quite as simply as "the Get API returns true". If we combined your proposal with keeping `AdditionalItemsReadyFunc` as an optional field where the generic solution fails, though, that might be workable.

Reply:
Another possible concern was that since most resource types don't actually need this waiting -- in most cases, the Kubernetes model of eventual consistency means that it's OK to restore the current item even if the `additionalItems` aren't available yet, so this would be needless waiting. Having the plugin signal when to wait makes this the exception rather than the rule. Since we weren't using a generic Get call for "is it ready?" we were able to combine "this plugin waits for additional items" with "this is how we wait" in one field. If we wanted to have the generic function be the default with `AdditionalItemsReadyFunc` as an override for how we wait, we could implement optional waiting with an additional boolean field here. That way we'd only wait when we're dealing with a resource which will fail restore if it's not yet there.  Then, once we have this (either as-is in the proposal, or with this new boolean field to allow for plugins not to have to redefine the method when the generic `Get` approach is good enough), we may not need the `--sync-restore-additional-items` option, although we could provide it -- if it's true, we wait for items when the plugin indicates we should, and if its' false, we never wait. But at this point, I don't know whether that option provides value. The point of having this in the first place is that with certain plugins/resources, not waiting will fail migrations, and with others waiting doesn't help. It feels like a plugin-specific setting rather than a restore-specific settings.

Reply:
@ashish-amarnath Just waiting for `GET` isn't quite enough for readiness all the time. Readiness means the thing exists and is useful.

As an example, CRDs can exist before CRDs can be made. Readiness is determined by examining `status.conditions`. [See this logic](https://github.com/vmware-tanzu/velero/blob/dbc83af77b0c0a36a27ac1646b6504fa78bedaba/pkg/util/kube/utils.go#L202-L209)

As another example, a `Pod` can exist but be in `CrashLoopBackOff`, and therefore not ready.

Reply:
Resolving this comment.